server/src/main/java/org/elasticsearch/action/search/SearchTaskStatus.java

```diff
@@ -0,0 +1,188 @@
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.elasticsearch.action.search;
+
+import org.elasticsearch.common.io.stream.StreamInput;
+import org.elasticsearch.common.io.stream.StreamOutput;
+import org.elasticsearch.common.io.stream.Writeable;
+import org.elasticsearch.common.xcontent.ToXContentFragment;
+import org.elasticsearch.common.xcontent.XContentBuilder;
+import org.elasticsearch.index.shard.ShardId;
+import org.elasticsearch.search.SearchPhaseResult;
+import org.elasticsearch.tasks.Task;
+import org.elasticsearch.tasks.TaskInfo;
+
+import java.io.IOException;
+import java.util.List;
+import java.util.concurrent.CopyOnWriteArrayList;
+import java.util.concurrent.atomic.AtomicReference;
+
+public class SearchTaskStatus implements Task.Status {
+
+    public static final String NAME = "search_task_status";
+
+    private final boolean readOnly;
+    private final List<PhaseInfo> phases;
+    private final AtomicReference<PhaseInfo> currentPhase;
+
+    SearchTaskStatus() {
+        this.phases = new CopyOnWriteArrayList<>();
+        this.currentPhase = new AtomicReference<>();
+        this.readOnly = false;
+    }
+
+    public SearchTaskStatus(StreamInput in) throws IOException {
+        this.currentPhase = new AtomicReference<>(in.readOptionalWriteable(PhaseInfo::new));
+        this.phases = in.readList(PhaseInfo::new);
+        this.readOnly = true;
+    }
+
+    @Override
+    public void writeTo(StreamOutput out) throws IOException {
+        out.writeOptionalWriteable(currentPhase.get());
+        out.writeList(phases);
+    }
+
+    @Override
+    public String getWriteableName() {
+        return NAME;
+    }
+
+    void phaseStarted(String phase, int expectedOps) {
+        assert readOnly == false;
+        boolean result = currentPhase.compareAndSet(null, new PhaseInfo(phase, expectedOps));
+        assert result : "another phase has not properly completed";
+    }
+
+    void phaseCompleted(String phase) {
+        assert readOnly == false;
+        PhaseInfo previous = currentPhase.getAndSet(null);
+        assert previous.name.equals(phase) : "current phase [" + previous.name +
+            "] is not the phase that's been completed [" + phase + "]";
+        phases.add(previous);
+    }
+
+    //TODO we should probably have a different method for failures, and be careful to count a shard only when all of its copies have failed
+
+    void shardProcessed(String phase, SearchPhaseResult searchPhaseResult) {
+        assert readOnly == false;
+        PhaseInfo phaseInfo = currentPhase.get();
+        assert phase.equals(phaseInfo.name) : "phase mismatch: current phase is [" + phaseInfo.name +
+            "] while shards are being reported processed for [" + phase + "]";
+
+        phaseInfo.shardProcessed(searchPhaseResult.getSearchShardTarget().getShardId(), searchPhaseResult.getTaskInfo());
```

Comment:
The task info can be null if the node is in a version that doesn't have the change ?